hazelcast/src/main/java/com/hazelcast/cache/impl/operation/OnJoinCacheOperation.java

```diff
@@ -59,7 +60,14 @@ public void run() throws Exception {
         if (isJCacheAvailable(getNodeEngine().getConfigClassLoader())) {
             ICacheService cacheService = getService();
             for (CacheConfig cacheConfig : configs) {
-                cacheService.putCacheConfigIfAbsent(cacheConfig);
+                // RU_COMPAT_38 since 3.9, configs are instances of OnJoinCacheConfig
+                CacheConfig cacheConfigToAdd;
+                if (cacheConfig instanceof OnJoinCacheConfig) {
+                    cacheConfigToAdd = ((OnJoinCacheConfig) cacheConfig).asCacheConfig();
```

Comment:
`OnJoinCacheConfig` should not leak outside `OnJoinCacheOperation`, so it is transformed to a `CacheConfig` as it is added to the `CacheService`'s config registry.